menas/src/main/scala/za/co/absa/enceladus/menas/services/SchemaRegistryService.scala

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2018 ABSA Group Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.enceladus.menas.services
+
+import java.net.URL
+
+import com.typesafe.config.ConfigFactory
+import org.springframework.beans.factory.annotation.{Autowired, Value}
+import org.springframework.stereotype.Service
+import za.co.absa.enceladus.menas.controllers.SchemaController
+import za.co.absa.enceladus.menas.models.rest.exceptions.RemoteSchemaRetrievalException
+import za.co.absa.enceladus.menas.services.SchemaRegistryService.SchemaResponse
+import za.co.absa.enceladus.menas.utils.SchemaType
+import za.co.absa.enceladus.utils.config.SecureConfig
+
+import scala.io.Source
+import scala.util.control.NonFatal
+
+@Service
+class SchemaRegistryService @Autowired()() {
+
+  @Value("${menas.schemaRegistryBaseUrl}")
+  val schemaRegistryBaseUrl: String = ""
+
+  private val config = ConfigFactory.load()
+  SecureConfig.setKeyStoreProperties(config)
+  SecureConfig.setTrustStoreProperties(config)
+
+  /**
+   * Loading the latest schema by a topicName (e.g. topic1-value), the url is constructed based on the [[schemaRegistryBaseUrl]]
+   * @param topicName topic name
+   * @return
+   */
+  def loadSchemaByTopicName(topicName: String): SchemaResponse = {
+    val schemaUrl = SchemaRegistryService.getLatestSchemaUrl(schemaRegistryBaseUrl, topicName)
+    loadSchemaByUrl(schemaUrl)
+  }
+
+  /**
+   * Loading the schema by a full URL
+   * @param remoteUrl
+   * @return
+   */
+  def loadSchemaByUrl(remoteUrl: String): SchemaResponse = {
+    try {
+      val url = new URL(remoteUrl)
+      val connection = url.openConnection()
+      val mimeType = SchemaController.avscContentType // only AVSC is expected to come from the schema registry
+      val fileStream = Source.fromInputStream(connection.getInputStream)
+      val fileContent = fileStream.mkString
+      fileStream.close()
+
+      SchemaResponse(fileContent, mimeType, url)
+    } catch {
+      case NonFatal(e) =>
+        throw RemoteSchemaRetrievalException(SchemaType.Avro, s"Could not retrieve a schema file from $remoteUrl. " +
```

Comment:
This `$remoteUrl` might show to the user more info than needed/wanted. Maybe show them just the topic?

Reply:
What kind of unwanted information do you imagine it can expose? The security here is done via the keystore/trustStore files (that do not get exposed). I do not expect any type of https://username:password@domain kind of configurations (this specific case could be addressed if we really wanted to)

Hmm, I actually find this way beneficial for the user. Given the fact that they can already put in the URL themselves, this error would give them insight into what URL has been tried and they can both check the SR if they have the means to do that (is the machine up? is the service up? is the path/topic-name correct?) and they can use this information to adjust the URL and use it in the "by URL remote loading" thereafter. The presence of the resulting URL can also help significantly with troubleshooting a deployed solution.

**I was actually thinking the opposite - that we might want to show them what the constructed URL (by the topic name stem) will be so that there is total visibility of what the tool aims to do.**

I see your point, but since we primarily aim to aid those using the tool in-house, more information available should be helpful, not hurtful.

--- 
One slight problem here is that the `loadSchemaByUrl` here is used for both cases - loading from URL and loading by topic-name stem, but that is the least of the concerns - that can I can easily adjust.

Can you confirm/refuse my "helpful-visibility-into-things" hypothesis, please?